### DISCUSSION TITLE
Using Target-Typed New with the params keyword

### DISCUSSION THREAD

**Discussion by brandon-irl** [dotnet/csharplang#4614] [Q&A] [answered]
Can someone tell me why I can't do this?

```csharp
        static void Main(string[] args)
        {
            DoSomething(new DateTime()); // This is fine
            DoSomething(new ());         // ERROR: The type 'DateTime[]' may not be used as the target type of new()
        }

        private static void DoSomething(params DateTime[] dates)
        {
            // TODO
        }
```

**Comment by 333fred** [comment, marked as answer]
Let me answer your question with a question: were you expecting that to give you a new array, or were you expecting that to give you a new `DateTime`? Because the compiler inferred the former (and has no choice but to infer the former), and to create a new array you need an array creation expression, not a constructor.

**Reply by brandon-irl** [reply]
Makes sense, but then  how does `DoSomething(new DateTime());` create a new array, and why can't `new()` do the same?

**Reply by 333fred** [reply]
> how does `DoSomething(new DateTime());` create a new array

This is, in general, how the `params` keyword words. When the compiler sees a params keyword, it allows you to _optionally_ put a series of expressions with the element type of the params array. `new DateTime()` has the same type as the element type of the params array, so it's allowed. `new()` doesn't have a type, it's inferred from context. When you provide a single element, the compiler has to figure out what you mean, and here `new()` could either refer to the element type of the params array, or it could refer to the array type itself. C# says that the compiler must _always_ prefer the params array type itself, not the element type, so `new()` is target-typed to `DateTime[]`, not `DateTime`.

**Reply by brandon-irl** [reply]
I see makes sense. I assumed that knowing that `new()` can't be used with array target types, the compiler would use the element type and not prefer the params array type. For anyone else reading this, `DoSomething(new (), new());` works and illustrates the answer.

**Reply by 333fred** [reply]
It is theoretically something we could change with a language proposal, I believe, but I'm not certain whether we explicitly decided against it during `new()` design. @alrz or @jcouv, do either of you remember if this fell out, or if it was an intentional decision?

**Reply by alrz** [reply]
Maybe the error is misleading. In a similar situation you get an ambiguous call, since `new()` does not contribute to overload resolution.

```cs
using System;
class C {   
    // effectively we have:
    private static void DoSomething(DateTime[] dates) {}
    private static void DoSomething(DateTime dates) {}  
    static void Main() {
        DoSomething(new()); // error CS0121: The call is ambiguous
    }
}
```
In params case, afaik we choose the expanded form based on the type of the arguments, but since `new()` doesn't have a type, it's resolved to the method as-is.
